File: src/core/timeline.ts
import type { OutputSegment, SessionEvent, Stroke } from "../types";

type State = { playing: boolean; sourceUs: number; rate: number; sessionUs: number };

export function buildTimeline(events: SessionEvent[]): OutputSegment[] {
  if (!events.length || events[0].type !== "session_start") return [];
  const start = events[0];
  const state: State = {
    playing: false,
    sourceUs: Number(start.payload.sourceUs ?? 0),
    rate: Number(start.payload.rate ?? 1),
    sessionUs: start.sessionUs,
  };
  const result: OutputSegment[] = [];

  for (const event of events.slice(1)) {
    const elapsed = Math.max(0, event.sessionUs - state.sessionUs);
    if (elapsed > 0) {
      if (state.playing) {
        const sourceEndUs = state.sourceUs + elapsed * state.rate;
        result.push({ kind: "play", outputStartUs: state.sessionUs, outputDurationUs: elapsed, sourceStartUs: state.sourceUs, sourceEndUs, rate: state.rate });
        state.sourceUs = sourceEndUs;
      } else {
        result.push({ kind: "freeze", outputStartUs: state.sessionUs, outputDurationUs: elapsed, sourceFrameUs: state.sourceUs });
      }
    }
    if (event.type === "play") {
      state.sourceUs = Number(event.payload.sourceUs ?? state.sourceUs);
      state.rate = Number(event.payload.rate ?? state.rate);
      state.playing = true;
    } else if (event.type === "pause") {
      state.sourceUs = Number(event.payload.sourceUs ?? state.sourceUs);
      state.playing = false;
    } else if (event.type === "seek") {
      state.sourceUs = Number(event.payload.toSourceUs ?? state.sourceUs);
      state.playing = Boolean(event.payload.continuesPlaying);
    } else if (event.type === "rate_change") {
      state.sourceUs = Number(event.payload.sourceUs ?? state.sourceUs);
      state.rate = Number(event.payload.toRate ?? state.rate);
      state.playing = Boolean(event.payload.playing);
    } else if (event.type === "session_stop" || event.type === "interruption") {
      state.playing = false;
    }
    state.sessionUs = event.sessionUs;
  }
  return mergeAdjacent(result);
}

function mergeAdjacent(segments: OutputSegment[]): OutputSegment[] {
  const merged: OutputSegment[] = [];
  for (const segment of segments) {
    const previous = merged.at(-1);
    if (previous?.kind === "play" && segment.kind === "play" && previous.rate === segment.rate && Math.abs(previous.sourceEndUs - segment.sourceStartUs) <= 2) {
      previous.outputDurationUs += segment.outputDurationUs;
      previous.sourceEndUs = segment.sourceEndUs;
    } else if (previous?.kind === "freeze" && segment.kind === "freeze" && previous.sourceFrameUs === segment.sourceFrameUs) {
      previous.outputDurationUs += segment.outputDurationUs;
    } else merged.push({ ...segment });
  }
  return merged;
}

export function sourceAtOutput(segments: OutputSegment[], outputUs: number): number {
  if (!segments.length) return 0;
  const match = segmentAtOutput(segments, outputUs);
  if (match) return match.kind === "freeze" ? match.sourceFrameUs : match.sourceStartUs + (outputUs - match.outputStartUs) * match.rate;
  const last = segments.at(-1)!;
  return last.kind === "freeze" ? last.sourceFrameUs : last.sourceEndUs;
}

export function segmentAtOutput(segments: OutputSegment[], outputUs: number): OutputSegment | undefined {
  let low = 0, high = segments.length - 1;
  while (low <= high) {
    const mid = (low + high) >>> 1;
    const current = segments[mid];
    if (outputUs < current.outputStartUs) high = mid - 1;
    else if (outputUs >= current.outputStartUs + current.outputDurationUs) low = mid + 1;
    else return current;
  }
  return undefined;
}

export function buildStrokes(events: SessionEvent[]): Stroke[] {
  const strokes = new Map<string, Stroke>();
  for (const event of events) {
    if (event.type === "stroke_begin") {
      const id = String(event.payload.strokeId);
      strokes.set(id, { id, color: String(event.payload.color), widthNormalized: Number(event.payload.widthNormalized), startUs: event.sessionUs, points: [{ x: Number(event.payload.x), y: Number(event.payload.y), sessionUs: event.sessionUs, pressure: 1 }] });
    } else if (event.type === "stroke_point" || event.type === "stroke_end") {
      const stroke = strokes.get(String(event.payload.strokeId));
      if (stroke) stroke.points.push({ x: Number(event.payload.x), y: Number(event.payload.y), sessionUs: event.sessionUs, pressure: Number(event.payload.pressure ?? 1) });
    } else if (event.type === "stroke_hide") {
      for (const id of event.payload.strokeIds as string[]) {
        const stroke = strokes.get(id);
        if (stroke && stroke.hiddenAtUs == null) stroke.hiddenAtUs = event.sessionUs;
      }
    }
  }
  return [...strokes.values()];
}

File: src-tauri/src/timeline.rs
use crate::model::{OutputSegment, SessionEvent};

pub fn build(events: &[SessionEvent]) -> Result<Vec<OutputSegment>, String> {
    let first = events.first().ok_or("イベントがありません。")?;
    if first.event_type != "session_start" {
        return Err("先頭イベントがsession_startではありません。".into());
    }
    let mut playing = false;
    let mut source_us = number(&first.payload, "sourceUs").unwrap_or(0) as f64;
    let mut rate = float(&first.payload, "rate").unwrap_or(1.0);
    let mut session_us = first.session_us;
    let mut segments = Vec::new();
    for event in events.iter().skip(1) {
        let elapsed = event.session_us.saturating_sub(session_us);
        if elapsed > 0 {
            if playing {
                let end = source_us + elapsed as f64 * rate;
                segments.push(OutputSegment::Play {
                    output_start_us: session_us,
                    output_duration_us: elapsed,
                    source_start_us: source_us.round() as u64,
                    source_end_us: end.round() as u64,
                    rate,
                });
                source_us = end;
            } else {
                segments.push(OutputSegment::Freeze {
                    output_start_us: session_us,
                    output_duration_us: elapsed,
                    source_frame_us: source_us.round() as u64,
                });
            }
        }
        match event.event_type.as_str() {
            "play" => {
                source_us = number(&event.payload, "sourceUs").unwrap_or(source_us as u64) as f64;
                rate = float(&event.payload, "rate").unwrap_or(rate);
                playing = true;
            }
            "pause" => {
                source_us = number(&event.payload, "sourceUs").unwrap_or(source_us as u64) as f64;
                playing = false;
            }
            "seek" => {
                source_us = number(&event.payload, "toSourceUs").unwrap_or(source_us as u64) as f64;
                playing = event
                    .payload
                    .get("continuesPlaying")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false);
            }
            "rate_change" => {
                source_us = number(&event.payload, "sourceUs").unwrap_or(source_us as u64) as f64;
                rate = float(&event.payload, "toRate").unwrap_or(rate);
                playing = event
                    .payload
                    .get("playing")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false);
            }
            "session_stop" | "interruption" => playing = false,
            _ => {}
        }
        session_us = event.session_us;
    }
    Ok(segments)
}

fn number(value: &serde_json::Value, key: &str) -> Option<u64> {
    value.get(key)?.as_u64()
}
fn float(value: &serde_json::Value, key: &str) -> Option<f64> {
    value.get(key)?.as_f64()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    fn e(seq: u64, session_us: u64, kind: &str, payload: serde_json::Value) -> SessionEvent {
        SessionEvent {
            schema_version: 1,
            seq,
            session_us,
            event_type: kind.into(),
            payload,
        }
    }
    #[test]
    fn play_pause_seek_rate_are_reconstructed() {
        let result = build(&[
            e(
                1,
                0,
                "session_start",
                json!({"sourceUs":2_000_000,"rate":1.0}),
            ),
            e(
                2,
                3_000_000,
                "play",
                json!({"sourceUs":2_000_000,"rate":1.0}),
            ),
            e(
                3,
                5_000_000,
                "seek",
                json!({"toSourceUs":0,"continuesPlaying":true}),
            ),
            e(
                4,
                7_000_000,
                "rate_change",
                json!({"sourceUs":2_000_000,"toRate":0.5,"playing":true}),
            ),
            e(5, 9_000_000, "session_stop", json!({})),
        ])
        .unwrap();
        assert_eq!(
            result[0],
            OutputSegment::Freeze {
                output_start_us: 0,
                output_duration_us: 3_000_000,
                source_frame_us: 2_000_000
            }
        );
        assert_eq!(
            result[3],
            OutputSegment::Play {
                output_start_us: 7_000_000,
                output_duration_us: 2_000_000,
                source_start_us: 2_000_000,
                source_end_us: 3_000_000,
                rate: 0.5
            }
        );
    }
    #[test]
    fn all_rates_calculate_source_duration() {
        for rate in [0.25, 0.5, 0.75, 1.0, 1.25, 1.5, 2.0] {
            let result = build(&[
                e(1, 0, "session_start", json!({"sourceUs":0,"rate":rate})),
                e(2, 0, "play", json!({"sourceUs":0,"rate":rate})),
                e(3, 4_000_000, "session_stop", json!({})),
            ])
            .unwrap();
            match result[0] {
                OutputSegment::Play { source_end_us, .. } => {
                    assert_eq!(source_end_us, (4_000_000.0 * rate) as u64)
                }
                _ => panic!(),
            }
        }
    }
}
